examples/connection_pools/src/main.rs

```diff
@@ -0,0 +1,176 @@
+#![feature(plugin, decl_macro)]
+#![plugin(rocket_codegen)]
+
+// TODO: Fix impls to extract multiple connection pools from state
+// As it stands now, only one database type can be fetched from state.
+
+extern crate rocket;
+extern crate rocket_contrib;
+
+use std::ops::Deref;
+use rocket::http::Status;
+use rocket::request::{self, FromRequest};
+use rocket::{Outcome, Request, State};
+use rocket::config;
+use rocket_contrib::databases::Poolable;
+use rocket_contrib::databases::diesel;
+use rocket_contrib::databases::postgres;
+use rocket_contrib::databases::r2d2;
+use rocket_contrib::databases::r2d2_postgres;
+
+// The two annotations would be part of the proc-macro crate to generate the
+// FromRequest implementation for the specified DB type.
+// #[derive(DbConn)]
+// #[database("my_sqlite_database")]
+struct TempStorage(r2d2::PooledConnection<diesel::r2d2::ConnectionManager<diesel::SqliteConnection>>);
+
+// #[derive(DbConn)]
+// #[database("my_primary_database")]
+struct PrimaryDb(r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>);
+
+// #[derive(DbConn)]
+// #[database("my_warehouse_database")]
+struct WarehouseDb(r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>);
+
+fn main() {
+    let rocket = rocket::ignite();
+    let config = rocket.config().clone();
+    rocket
+        .mount("/", routes![test_handler])
+        .manage(TempStorage::pool(config.clone()))
+        .manage(PrimaryDb::pool(config.clone()))
+        .manage(WarehouseDb::pool(config))
+        .launch();
+}
+
+#[get("/")]
+fn test_handler(_storage: TempStorage, _primary: PrimaryDb, _warehouse: WarehouseDb) -> &'static str {
+    "Hello, Rocket with databases!"
+}
+
+// Code below here would be generated at compile-time via procedural macros
+struct TempStoragePool(r2d2::Pool<<diesel::SqliteConnection as Poolable>::Manager>);
+
+impl TempStorage {
+    fn pool(config: rocket::Config) -> TempStoragePool {
+        let db_config: config::Table = config
+            .get_slice("databases")
+            .unwrap()
+            .iter()
+            .filter(|element| element.as_table().unwrap().get("name").unwrap().as_str().unwrap() == "my_sqlite_db")
+            .nth(0)
+            .unwrap()
+            .as_table()
+            .unwrap()
+            .clone();
```

Comment:
This is ugly, repeated three times, and not elegant. While it isn't code the user would write, it's not something I think we want to code gen either. Just seemed like the best way to get what I needed for now.

Reply:
My guess is that we should provide a free function (or something) in the `databases` library that return the configuration for a given database by name. Something like:

```rust
fn database_config(name: &str, from: &Config) -> Option<&Table> { .. }
```

See my other comment wondering if we can make these types more type-safe.

Reply:
I like that idea. Smart to provide the user a method to retrieve the configuration for them.

Reply:
My guess is that the user won't use this method very much, but it'll be less code we need to generate. In general, we should err on the side of generating a little code as possible since none of it type-checked until it's generated.

Reply:
👍